Fixes #12098 - Dangling Fog.mock makes compute_resource_vms test fail

dLobatog opened the pull request:
test/unit/compute_resource_test.rb contains a Fog.mock! without a
Fog.unmock!. This causes
test/functional/compute_resource_vms_controller_test.rb to fail only
when ran through rake test on Rails 4. Fix is as simple as adding the
missing unmock!.

shlomizadok commented:
Maybe move it to teardown so future tests could use it too?

dLobatog commented:
@shlomizadok I could put Fog.mock! and unmock in setup and teardown, but it seems there's only one test using it, so we would be calling it without any purpose on other tests

shlomizadok commented:
@eLobato - right. However, if you only puts it in teardown it will ensure that each test that will be written in the future will unmock fog.
Please also see in [openstack test](https://github.com/theforeman/foreman/blob/cf043905057d77d00618d9b0c8c3c21b51a9d425/test/unit/compute_resources/openstack_test.rb#L22) and [it's teardown](https://github.com/theforeman/foreman/blob/cf043905057d77d00618d9b0c8c3c21b51a9d425/test/unit/compute_resources/openstack_test.rb#L11-L13)

domcleal commented:
I'd agree with moving it to a teardown instead of inside the test method itself, as it means the unmock will get called if an assertion fails - and doesn't leave Fog in an inconsistent state.  (Though arguably, we shouldn't be using unmocked calls, but I think we have some.)

Perhaps we don't have to do it for every test - define a method in test_helper, in one of the test case parent classes, and then tests who need it could just add `teardown :unmock_fog`?

dLobatog commented:
@domcleal How do you check which tests need `Fog.unmock!`?

domcleal commented:
Just the ones we know about, like the ones in this file.

eLobato commented:
There were a lot of calls to Fog.mock and unmock that were unwarranted before. This last commit ensures setup

dLobatog committed:
Fixes #12098 - Dangling Fog.mock! makes compute_resource_vms test fail

test/unit/compute_resource_test.rb contains a Fog.mock! without a
Fog.unmock!. This causes
test/functional/compute_resource_vms_controller_test.rb to fail only
when ran through rake test on Rails 4. Fix is as simple as adding the
missing unmock!.

dLobatog commented:
@domcleal Is this what you were looking for? I could define a method in the helper but I think the `teardown(:method)` syntax doesn't work within tests.

dLobatog commented:
[test]

domcleal commented:
It's not quite - I was hoping we could define a fog_unmock in test_helper, then at the top of this test file (not inside the test method), put `teardown :fog_unmock`.  It blew up spectacularly just now when I tried it, causing the test process to OOM.  I think I'll live with this!

domcleal commented:
Merged as f8a45eb, thanks @dLobatog.
